samza-core/src/main/java/org/apache/samza/operators/impl/InputOperatorImpl.java

```diff
@@ -35,13 +35,12 @@
  *
  * @param <K> the type of key in the incoming message
  * @param <V> the type of message in the incoming message
- * @param <M> the type of input message
  */
-public final class InputOperatorImpl<K, V, M> extends OperatorImpl<Pair<K, V>, M> {
+public final class InputOperatorImpl<K, V> extends OperatorImpl<KV<K, V>, Object> { // Object == KV<K,V> | V
```

Comment:
Is there another InputOperator* for unkeyed inputs?

Reply:
Not really. The messages we get as an input to this operator are always keyed, with the key being that of InputMessageEnvelope. This operator is where we decide whether to propagate the key further, based on whether the input stream Serde is keyed or not.